tests/irdl/test_range_constraint.py

```diff
@@ -44,3 +48,58 @@ def test_range_of_variables():
     attr_constr = VarConstraint("ATTR", AnyAttr())
     len_constr = IntVarConstraint("LENGTH", AnyInt())
     assert RangeOf(attr_constr, length=len_constr).variables() == {"ATTR", "LENGTH"}
+
+
+def test_verify_correct_length():
+    hello = StringAttr("hello")
+    world = StringAttr("world")
+    attr_constr = VarConstraint("ATTR", BaseAttr(StringAttr))
+    len_constr = IntVarConstraint("LENGTH", AnyInt())
+    range_len_constr = RangeLengthConstraint(RangeOf(attr_constr), len_constr)
+    with pytest.raises(
+        VerifyException,
+        match='''attribute "hello" expected from variable 'ATTR', but got "world"''',
+    ):
+        range_len_constr.verify((hello, world), ConstraintContext())
+    range_len_constr.verify((world, world), ConstraintContext())
+
+    with pytest.raises(
+        VerifyException,
+        match="incorrect length for range variable",
+    ):
+        range_len_constr.verify(
+            (world, world, world), ConstraintContext(_int_variables={"LENGTH": 2})
+        )
+
+    # verify_length
+    range_len_constr.verify_length(2, ConstraintContext())
+
+    with pytest.raises(
+        VerifyException,
+        match="incorrect length for range variable",
+    ):
+        range_len_constr.verify_length(
+            3, ConstraintContext(_int_variables={"LENGTH": 2})
+        )
+
+    # variables
+    assert range_len_constr.variables() == {"ATTR", "LENGTH"}
+
+    # can_infer
+    assert not range_len_constr.can_infer(set(), length_known=True)
+    assert not range_len_constr.can_infer(set(), length_known=False)
+    assert not range_len_constr.can_infer({"ATTR"}, length_known=False)
+    assert range_len_constr.can_infer({"ATTR"}, length_known=True)
+    assert range_len_constr.can_infer({"ATTR", "LENGTH"}, length_known=False)
+
+    # infer
+    assert range_len_constr.infer(
+        ConstraintContext(_variables={"ATTR": world}), length=2
+    ) == (
+        world,
+        world,
+    )
```

Comment:
One thing I'm not sure about is whether inference should verify the length here, feels like it would be a bit weird if we could infer a range of an unsupported length. This would be a change on the test, and the implementation of infer would have to call verify in the case where `length` is not `None`.
```suggestion
    ctx = ConstraintContext(_variables={"ATTR": world})
    assert range_len_constr.infer(ctx, length=2) == (
        world,
        world,
    )
    assert ctx.get_int_variable("LENGTH") == 2
```

Reply:
I don't see why this would be necessary, as far as I know, we don't require inference to verify things (unless required by pyright typechecking usually).